Moore/multisample

timoore opened the pull request:
# Pull Request Template

## Description

This implements the creation of a multisampled frame buffer for vsg::Viewer. If the new WindowTraits::samples member requests more than 1 bit of sampling, then a render pass and frame buffer are created with a multisampled attachment that automatically resolves, via subpass flags, to the swapchain attachment.

A FrameAssembly class has been added to abstract the interaction between the RenderGraph and a window or render-to-texture frame buffer. If you like the approach, I might add more convenience functions to that class with the objective of supporting multisampling render-to-texture.

An update to vsgviewer in vsgExamples is being submitted as well.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

If multisampling is requested, then the scene graph needs to be modified with the proper multisampling state.

## How Has This Been Tested?

I ran vsgviewer, both with sampling and without

- [X] Test A
vsgviewer --samples 4 teapot.vsgt
- [ ] Test B

**Test Configuration**:
* Firmware version:
* Hardware: GTX 1070
* Toolchain: Linux
* SDK:1.2.131.2

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

robertosfield commented:
Great to the addition of multi-sampling support, I am taken aback by just how much files have been changed to provide this functionality.  The changes are too extensive for me to be able to just slot directly into the VSG without a deep dive into the topic and the approach taken in this implementation.

What I'll do is create a multisample branch, we can then use that to merge the functionality for wider review.

My plan is to do review of subpass and secondary command buffers next week, There is some overlap between the files that will changed with the ones change in adding multi-pass support so it'll be worth me doing the reviews somewhat in parallel.

One of the things I'm keen to do is mirror language used in the underlying Vulkan implementation as our classes that encapsulate that functionality.  Where we introduce new language we'll need to make sure that it provides clarity in how it fits into the wider application and scene graph.  I don't know where we'll end up on this, but it's something I'm keeping in mind as I merged changes or make changes.

I'll have a chance to make the multisample branch in the next hour when I get back to my dev system.

robertosfield commented:
I have created a multisample branch on vsg-dev/VulkanSceneGraph and merged this PR into it.

    https://github.com/vsg-dev/VulkanSceneGraph/tree/multisample

I'll do the same for the PR to vsgExamples.  

Next week I'll dive into the topic and figure out a route through.

timoore commented:
I agree with putting this in a branch for now. If I make a path to support multisampled render-to-texture, I'll create a pull request against that branch.

I also agree about using Vulkan vocabulary for concepts, classes and functions in VSG. I kind of made up "FrameAssembly" because its something that a user would think of as a frame buffer, but a Vulkan FrameBuffer is something much more limited.

robertosfield commented:
Naming is hard. Can take several iterations to get it right.

If needs be we can create more branches to experiment with various options.

On Sat, 16 May 2020 at 14:32, timoore <notifications@github.com> wrote:

> I agree with putting this in a branch for now. If I make a path to support
> multisampled render-to-texture, I'll create a pull request against that
> branch.
>
> I also agree about using Vulkan vocabulary for concepts, classes and
> functions in VSG. I kind of made up "FrameAssembly" because its something
> that a user would think of as a frame buffer, but a Vulkan FrameBuffer is
> something much more limited.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/vsg-dev/VulkanSceneGraph/pull/179#issuecomment-629646320>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEGUAPXHVH2FYL72B7BODRR2IWFANCNFSM4NC4BP4Q>
> .
>

robertosfield commented:
@timoore phew, just walked the narrow ledge whilst juggling to pull the more general purpose changes in the multismaple branch into master, and then merged the changes back into multisample and multisample_alternate) so that these two multisample branches just contain the parts that a very specific to the multi-sampling support.  I "think" it's all merged back and forth correctly, with both master and the multisample branches working fine, would be worth being mindful that regressions might be introduced with me doing the juggling of code.

At this point the multisample and multisample_alternate branches are the same code wise.

I have also got most of the key changes in the vsgExamples multisample branches merged into master.

If you are planning any further changes yourself you'll need to pull from VSG and vsgExmaple repo's multisample branch.

I will now take a breather from review all this code and when I come back I'll see if I can spot a cleaner alternative to the FrameAssembly functionality.  It might be that the FrameAssembly is right approach, and just by looking into it all more deeply this might be the final approach to be merged with master.  Any experiments I'll multisample_alternate branch so we compare like for like.

robertosfield commented:
@timoore Just getting back to reviewing the multipass branch.  Could you explain the motivation behind FrameAssembly as it's not yet clear to me what problem it solves and how it's intended to be used. If not obvious to me then I'm sure others following along behind will be similarly unsure about what it's all about.

robertosfield commented:
@timoore I've decided rather think about refactoring FrameAssembly to something different I would just learn what it takes at the Vulkan level to support multi-sampling bare bones without any extra classes.  

The first step has been to reset the multisample_alternate branch to master, so it now is a none functioning multisample branch.  As I pulled what I could from the multisample branch into master I'm not starting from scratch.  All the goodness in Context/GraphicsPipeline etc. is all there.  What I'm missing is all the multi-sampling glue inside RenderGraph/Window and FrameAssembly.  

I am beginning to suspect FameAssembly is there to solve more than the problem of multi-sampling, so I'm still curious about what problems it might provide solutions too.  Essentially I'm interesting in the problems that need resolving rather than the solution.  I think this is kinda orthogonal to the task of supporting multi-sampling, hence my decision to attempt to build multi-sampling support from master.  Even if this new multisample_alternate path goes nowhere I'll still have learned something :-)

timoore commented:
Hi,
Originally, FrameAssembly was written to remove a bunch of special-case
code inside RenderGraph for the support of offscreen rendering. It is
simply the aggregate of a render pass and a framebuffer which is needed to
actually do rendering. It turns out that the attachments may change, like
for a Window, but that's OK because the FrameAssembly just returns a render
pass and framebuffer that is valid for the current frame. The rest of VSG
doesn't care if those are associated with a surface somehow are are backed
entirely by images.

Multisample requires a sightly more complicated attachment and subpass
structure, and different clear values, but again the details can be hidden
behind FrameAssembly. Render passes and framebuffers are always going to be
tightly bound together; therefore, I think that the FrameAssembly class has
a lot of potential for managing the creation of subpasses and attachments.

Tim

On Wed, May 20, 2020 at 6:45 PM Robert Osfield <notifications@github.com>
wrote:

> @timoore <https://github.com/timoore> I've decided rather think about
> refactoring FrameAssembly to something different I would just learn what it
> takes at the Vulkan level to support multi-sampling bare bones without any
> extra classes.
>
> The first step has been to reset the multisample_alternate branch to
> master, so it now is a none functioning multisample branch. As I pulled
> what I could from the multisample branch into master I'm not starting from
> scratch. All the goodness in Context/GraphicsPipeline etc. is all there.
> What I'm missing is all the multi-sampling glue inside RenderGraph/Window
> and FrameAssembly.
>
> I am beginning to suspect FameAssembly is there to solve more than the
> problem of multi-sampling, so I'm still curious about what problems it
> might provide solutions too. Essentially I'm interesting in the problems
> that need resolving rather than the solution. I think this is kinda
> orthogonal to the task of supporting multi-sampling, hence my decision to
> attempt to build multi-sampling support from master. Even if this new
> multisample_alternate path goes nowhere I'll still have learned something
> :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vsg-dev/VulkanSceneGraph/pull/179#issuecomment-631593525>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABAQAPLXJNKLWRRGVKB4G3RSQCLFANCNFSM4NC4BP4Q>
> .
>

robertosfield commented:
Thanks for the explanation.

robertosfield commented:
Hi @timoore, 

I have now got the multsample_alternate branch working so functionally it's identical to the multsample branch, the only difference between the two is the use of FrameAssumbly.

As I've done this work and been using the original multsample branch as reference it feels like FrameAssembly is reinventing some of the RenderGraph functionality, filling in gaps where it's not yet flexible enough, in particular decoupling from the explicit Window dependency.  This takes us back to the previous changes you had to make to RenderGraph that again shone a light on the deficiencies of RenderGraph as it stood.

My current inclination is towards fixing RenderGraph so it's more flexible, decoupling from the explict coupling with Window in a more wholesale way than with your earlier changes for the render to texture work.

What I think it worth doing next is look at refactoring RenderGraph, and I'll start that after lunch.  I don't think it need be a massive change to how one interfaces with it and how it works internally.

On the multisampling front I'll do a another review of the multisample_alternate and multisample branches to see if I've missed anything.  The multisample_alternate is lighter-weight version that achieves the same functionality so I think it's a better candidate for merging with master.

For yourself I think it would be worth reviewing the multisample_alternate branch, partly to spot anything missing in what I implemented, and partly to see how I took elements of your work and introduced in ways that are simplifications of it.  Expressing a solution to a problem in the simplest way is one of the driving forces for my work on the VSG, it's one of the big lessons from working on the OSG over the years, too often I merged complicated code that made the code harder and harder to maintain.  Simpler code that fulfills the requirements is harder to write initially but long term it's far easier to maintain.

robertosfield commented:
@timoore I have done another code review and pretty happy with the multisample_alternate branch, it only adds another 105 lines of code vs the 228 added by the multsample brach, and doesn't added any new levels of indirection that the FrameAssembly does.

I will leave the multisample branch up for the time being as we can use it as a reference for a refactor of RenderGraph.

timoore commented:
I'll start taking a look. I agree that RenderGraph should be decoupled from
Window; there's absolutely nothing special about a window that's
relevant to RenderGraph. Even the changing extent is something that
offscreen rendering should handle. To my mind RenderGraph has one job:
render a graph in a VkRenderPass. The Vulkan RenderPass and Framebuffer are
tightly coupled, although not always. Should they be tied in the
RenderGraph, or in another object that' closer to the Vulkan metal?

Tim

On Thu, May 21, 2020 at 12:58 PM Robert Osfield <notifications@github.com>
wrote:

> Hi @timoore <https://github.com/timoore>,
>
> I have now got the multsample_alternate branch working so functionally
> it's identical to the multsample branch, the only difference between the
> two is the use of FrameAssumbly.
>
> As I've done this work and been using the original multsample branch as
> reference it feels like FrameAssembly is reinventing some of the
> RenderGraph functionality, filling in gaps where it's not yet flexible
> enough, in particular decoupling from the explicit Window dependency. This
> takes us back to the previous changes you had to make to RenderGraph that
> again shone a light on the deficiencies of RenderGraph as it stood.
>
> My current inclination is towards fixing RenderGraph so it's more
> flexible, decoupling from the explict coupling with Window in a more
> wholesale way than with your earlier changes for the render to texture work.
>
> What I think it worth doing next is look at refactoring RenderGraph, and
> I'll start that after lunch. I don't think it need be a massive change to
> how one interfaces with it and how it works internally.
>
> On the multisampling front I'll do a another review of the
> multisample_alternate and multisample branches to see if I've missed
> anything. The multisample_alternate is lighter-weight version that achieves
> the same functionality so I think it's a better candidate for merging with
> master.
>
> For yourself I think it would be worth reviewing the multisample_alternate
> branch, partly to spot anything missing in what I implemented, and partly
> to see how I took elements of your work and introduced in ways that are
> simplifications of it. Expressing a solution to a problem in the simplest
> way is one of the driving forces for my work on the VSG, it's one of the
> big lessons from working on the OSG over the years, too often I merged
> complicated code that made the code harder and harder to maintain. Simpler
> code that fulfills the requirements is harder to write initially but long
> term it's far easier to maintain.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vsg-dev/VulkanSceneGraph/pull/179#issuecomment-632022189>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABAQAIK7UQR2KBDRLUVNADRSUCL7ANCNFSM4NC4BP4Q>
> .
>

robertosfield commented:
On Thu, 21 May 2020 at 14:18, timoore <notifications@github.com> wrote:

> I'll start taking a look. I agree that RenderGraph should be decoupled from
> Window; there's absolutely nothing special about a window that's
> relevant to RenderGraph. Even the changing extent is something that
> offscreen rendering should handle. To my mind RenderGraph has one job:
> render a graph in a VkRenderPass. The Vulkan RenderPass and Framebuffer are
> tightly coupled, although not always. Should they be tied in the
> RenderGraph, or in another object that' closer to the Vulkan metal?
>

I want to avoid having multiple levels of indirection, we already have
RenderPass -> RenderPass::Implementation ->VkRenderPass, so I really don't
want to add more.

RenderGraph is a about pulling together all the high level objects required
to do graphics rendering, it's an object that you configure once and then
have it be actively used there after without much need to querying it's
current state, so it's configure at start up and pretty well forget for the
rest of the lifetime of the app.

The configuration of the RenderGraph to suite rendering to a window or an
off screen buffer is something that once done at startup, having a couple
of help functions do this configuration for us should be fine.

RenderGraph isn't quite there yet though... so it just need to go through a
bit further distillation.   If it turns out that it can't be adapted to fit
the requirements then we can look at adding another layer of
RenderPass/Framebuffer management, I'd much rather avoid adding more
classes as it's just more code for users to understand and more code for us
to debug and maintain.

robertosfield commented:
@timoore as a heads up.  To help explore the deficinces of the present RenderGraph I've updated the vsgrendertorexture to make the creation of RenderGraph and CommandGraph clearer and more explicit.

https://github.com/vsg-dev/vsgExamples/commit/5b31de119ce7a06cd66d03417682678340abb6cd

The RenderGraph setup side looks pretty clean to me, the helper function to create the RendeGraph to texture could possibly do with some convenience methods in the core VSG to help out, but I'm pretty happy at the basic level of the present interface.

Internally it's less ideal so I'll now focus on the "if (window)" blocks in CompileTraversal and RenderGraph.

robertosfield commented:
@timoore A quick heads up.  I've done a bit of work on a new RenderGraph_refinement branch, in particular looking at ways to decouple RenderGraph from Window so it just works with RenderPass and Framebuffers.  I can see a route through, but getting there will require some changes beyond RenderGraph.

One of the tweaks I made was to add a RenderPass::maxSamples() method that returns the high attachment samples value and  use this in the CompileTraversal to help do part of the decoupling.

    https://github.com/vsg-dev/VulkanSceneGraph/compare/RenderGraph_refinement

I also fixed a resize issue that occurred during multi-sampling, which I also cherry-picked into master.  

A made further changes to experiment a little.  I have a good enough feel that more can be done with RenderGraph to decouple it from Window but haven't get a strong feel for any particular solution.  Rather than force some conclusion on this matter I've decided to just let the topic sit for a few days/week and come back to it.  

Tomorrow, I'm going to move back to considering the topic subpasses, secondary command buffers and execute command.  That work might have some input on the RenderGraph_refinement work.

robertosfield commented:
@timoore FYI, I've renamed the original multisample branch to FrameAssembly in VSG and vsgExamples as it's main focus/unique aspect is now the FrameAssembly rather than muiltisampling.  

I'm keen to avoid possible confusion with end users as multisampling is now part of master.
